### PR TITLE
Removed `class_name` cargo culting

### PR DESCRIPTION
**Pull request by jamesbraza**
# Description

When making a new subclass, I discovered this cargo culting of `class_name`, and I removed it.

## Type of Change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

- [ ] Added new unit/integration tests
- [ ] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

### CODE REVIEW

File: llama_index/schema.py [hunk 1]
```diff
@@ -23,9 +23,9 @@ class BaseComponent(BaseModel):
     """Base component object to caputure class names."""
 
     @classmethod
-    @abstractmethod
     def class_name(cls) -> str:
         """Get class name."""
+        return cls.__name__
```
Comment (jamesbraza) [inline llama_index/schema.py:28]: The universal solution, will work in all cases regardless of subclassing

Comment (logan-markewich) [reply llama_index/schema.py:28]: @jamesbraza actually, this needs to be a hard-coded string. It's for serialization purposes, so if the class name changes, we don't want to break a bunch of stuff

Comment (logan-markewich) [reply llama_index/schema.py:28]: We have some other work to rename this attribute. But for now, this is just how it is

Comment (jamesbraza) [reply llama_index/schema.py:28]: Fully qualified name changes when class name changes, so imo a changed class name _should_ break serialization.

Can you describe the serialization purposes?  I would then like to add the justification to the base class's `class_name` docstring.

Note in my deleting some of these, there's errors in a few of them

Comment (jamesbraza) [reply llama_index/schema.py:28]: If you're using the keys from `to_dict` elsewhere, then I guess yeah feel free to close this PR

**Comment by jamesbraza** [comment]
Gonna close this out

File: llama_index/schema.py (hunk 1 above)
Comment (logan-markewich) [reply llama_index/schema.py:28]: Yea so, we use pydantic to serialize, and have a consistent class name/literal works well as a discriminator when loading the class (pydantic does not make loading serialized classes easy when there's a bunch of subclasses etc., gotta work around it)

Comment (jamesbraza) [reply llama_index/schema.py:28]: Yeah I know what you mean.  A good solution is to use fully-qualified path as a key.  Then it's unique, is tolerant to class name changes, and is a path to use for deserialization.

---

Also, reflecting on the `cls.__name__` solution posed in this PR, while working with current serialization requirements.  I guess if subclassers changed their class name, they'd then have to override `class_name` to match the old class's name.

Anyways, just food for thought, if you're redesigning this in the future

**Pull request closed without merging**